### ISSUE TITLE
COPY Pipeline

### ISSUE THREAD

**Issue by ray6080** [kuzudb/kuzu#1640]
This issue is to document our designs to integrate `COPY` with our processor pipelines.
*This is a draft now to be further discussed and more details need to be added.*

## Overview
The pipeline should be flexible enough to cover both node/rel property graph tables, and rdf graphs.
Also, the new design should make COPY generally efficient and more robust on large-scale datasets. Here, by robustness, I mean the system doesn't crash due to limited memory resources, and the performance of copy doesn't degrade too much when memory is not enough (involving spill to disk during loading).
Changes to COPY need to be in tandem with our new storage design changes, as we're changing underlying storage layout a lot.
So this design will be broken into steps to be gradually done. Hopefully we can get most done when we've moved to the new storage layout.

## Pipeline
The source operator of the COPY pipeline is always `READER`.
The `READER` operator takes a morsel ( a chunk of a file) as input, and outputs a `DataChunk`. Internally, it parses a chunk from input files (e.g., some num of lines from a csv file, a row group from a parquet file, etc.) and converts the chunk into a `DataChunk`, which can be passed along operators.
The conversion handles the mismatch between arrow arrays and ValueVectors, and should be zero-copy when possible. For example, converting a INT64 arrow array to a ValueVector requires copying of null mask, but can redirect pointers in ValueVector to the arrow array data.
Note: we will start a project on the side to replace arrow with our own csv and parquet readers, which will be described in a separate issue.

Multiple `READER` objects (in different threads) in the same pipeline share a `ReaderSharedState`. Same as any other source operators, such as `ScanNodeID`, `ReaderSharedState` is used to coordinate tasks in the pipeline. 
The shared state should cover different file formats, such as CSV, Parquet, NPY, etc. and also different data sources, such as local fs, cloud fs, python objects, etc.
`READER` operates in two modes: *SERIAL*, and *NON-SERIAL*. Currently, only copying of node table with serial data type as the primary key requires serial mode, as it requires that node tuples are given offsets in the same order as they are stored in files. In the serial mode, the READER will guarantee that it reads sequentially, and also reads a whole node group at one time, so each `COPY_NODE` object will pull the whole node group and process it without coordinating with other threads.
While in the non-serial mode, `READER` outputs one data chunk (2048 tuples) at one time, and each `COPY_NODE` object will merge the data chunk into its own current node group without knowing tuples' positions in files.

### Node copy pipeline
Node copy will be in two pipelines:
```
     COPY_SINK
         |
    BUILD_INDEX ----
                   | sharedState: primary key column
      COPY_NODE ----
         |
       READER
```

The node copy pipeline should avoid reading the original csv files twice (thus, we can't rely on the first pass reading to figure out num of rows, which hash index requires for bulk reserving), and also separate index building, which can be reusable in the future for `CREATE INDEX ...`.
To avoid flushing the primary key column to disk and read again, we need to introduce a caching mechanism, that we always keep the primary key column in RAM before we start to construct the index (assuming the column won't be too large, if it's too large, we should still flush to disk, and scan from disk sequentially again, which should be fine, too).

### Rel copy
Rel copy process is also broken into two pipelines (suppose it's a `MANY_ONE` relationship, that's why it contains both `COPY_REL_COLUMNS` and `COPY_REL_LISTS`):
```
# pipeline 2 (parent)
      COPY_SINK
          |
    COPY_REL_LISTS
          |
    COPY_REL_COLUMNS
          |
    DATA_CHUNK_SCAN   --
                        | sharedState: shuffled data chunks and tuple pointers for each node group
# piepline 1 (child)    |
     REL_SHUFFLE      --
          |
     INDEX_LOOKUP
          |
        READER
```
Pipeline 1 reads input files in data chunks (same as node copy), and perform index lookups on both src and dst node keys to get their internal node offsets. Based on src and dst node offsets, `REL_SHUFFLE` partitions each tuple into corresponding partitions (i.e., node groups). Note that we duplicate rel data in both forward and backward directions.
`REL_SHUFFLE` is a sink operator, it accumulates all data chunks read from source in memory, and for each node group, it records logical pointers (might be something similar to tuple idxes) to tuples in accumulated data chunks. Accumulated data chunks and logical pointers will be stored inside the shared state between `REL_SHUFFLE` and `DATA_CHUNK_SCAN`.

Pipeline 2 scans tuples per node group from the shared state, and populate columns and rel lists accordingly.

### RDF copy
TODO(Guodong): For rdf graphs, we should combine node copy and rel copy pipelines to avoid reading original rdf files twice (one for node table and one for rel table). Need to discuss with @gaurav8297 .
```
# pipeline 2 (parent)
      COPY_SINK
          |
    COPY_REL_LISTS
          |
     COPY_COLUMNS
          |
    DATA_CHUNK_SCAN   --
                        | sharedState: shuffled data chunks and tuple pointers for each node group
# piepline 1 (child)    |
     REL_SHUFFLE      --
          |
    BUILD_INDEX
          |
    COPY_COLUMNS
          |
       READER
```

## TODOs
Just a list of stuff in my mind here. Not ordered here. Need to be improved later.
- [x] Move csv and parquet node copier to operators.
- [x] Move NPY copier to operators.
- [ ] Separate hash index build and column copy.
- [x] Move csv and parquet rel copier to COPY pipeline.
- [x] Node group based node copy pipeline.
- [x] Node group based rel copy pipeline.
- [x] RDF copy pipeline.
